app/scripts/migrations/135.ts

```diff
@@ -0,0 +1,318 @@
+        if (rpcEndpoint.networkClientId === networkClientId) {
+          return chainId;
+        }
+      }
+    }
+
+    const builtInChainId = BUILT_IN_NETWORKS.get(networkClientId);
+    if (!builtInChainId) {
+      global.sentry?.captureException(
+        new Error(
+          `Migration ${version}: No chainId found for ${propertyName} "${networkClientId}"`,
+        ),
+      );
+    }
+    return builtInChainId;
+  };
+
+  const currentChainId = getChainIdForNetworkClientId(
+    selectedNetworkClientId,
+    'selectedNetworkClientId',
+  );
+  if (!currentChainId) {
+    return state;
+  }
+
+  for (const [origin, subject] of Object.entries(subjects)) {
+    if (!isObject(subject)) {
+      global.sentry?.captureException(
+        new Error(
+          `Migration ${version}: Invalid subject for origin "${origin}" of type ${typeof subject}`,
+        ),
+      );
+      return state;
+    }
+
+    const { permissions } = subject as {
+      permissions: Record<string, PermissionConstraint>;
+    };
+    if (!isObject(permissions)) {
+      global.sentry?.captureException(
+        new Error(
+          `Migration ${version}: Invalid permissions for origin "${origin}" of type ${typeof permissions}`,
+        ),
+      );
+      return state;
+    }
+
+    let basePermission: PermissionConstraint | undefined;
+
+    let ethAccounts: string[] = [];
+    if (
+      isObject(permissions[PermissionNames.eth_accounts]) &&
+      Array.isArray(permissions[PermissionNames.eth_accounts].caveats)
+    ) {
+      ethAccounts =
+        (permissions[PermissionNames.eth_accounts].caveats?.[0]
+          ?.value as string[]) ?? [];
+      basePermission = permissions[PermissionNames.eth_accounts];
+    }
+    delete permissions[PermissionNames.eth_accounts];
+
+    let chainIds: string[] = [];
+    if (
+      isObject(permissions[PermissionNames.permittedChains]) &&
+      Array.isArray(permissions[PermissionNames.permittedChains].caveats)
+    ) {
+      chainIds =
+        (permissions[PermissionNames.permittedChains].caveats?.[0]
+          ?.value as string[]) ?? [];
+      basePermission ??= permissions[PermissionNames.permittedChains];
+    }
+    delete permissions[PermissionNames.permittedChains];
+
+    if (ethAccounts.length === 0 || !basePermission) {
+      continue;
+    }
+
+    if (chainIds.length === 0) {
+      chainIds = [currentChainId];
+
+      const networkClientIdForOrigin = domains[origin];
+      if (
+        networkClientIdForOrigin &&
+        typeof networkClientIdForOrigin === 'string'
+      ) {
+        const chainIdForOrigin = getChainIdForNetworkClientId(
+          networkClientIdForOrigin,
+          'networkClientIdForOrigin',
+        );
+        if (chainIdForOrigin) {
+          chainIds = [chainIdForOrigin];
+        }
+      }
+    }
+
+    const isSnap = snapsPrefixes.some((prefix) => origin.startsWith(prefix));
+    const scopes: InternalScopesObject = {};
+    const scopeStrings: CaipChainId[] = isSnap
```

Comment:
I don't entirely follow the reasoning here, what about Snaps that have accounts created?

Reply:
They'll still be stored under the `wallet:eip155` scope a few lines below this

Reply:
Hmm, I don't follow that either, what is that scope supposed to represent?

Reply:
that scope represents evm wallet methods, i.e. methods that do not belong to a particular evm chain, so something like `personal_sign` would belong here . It's kind of awkward right now because our current Signing UI implementation does require chainId context and so some signing methods do belong to a chain specific scope (i.e. `eip155:1` etc) and so `personal_sign` does not exist in `wallet:eip155` like in my example above, but I won't go into the specifics there

Reply:
basically, it's a scope for evm stuff, but not chain specific

Reply:
So given that Snaps may already have eth_accounts permissions on technically all chains, why aren't we granting that permission in the same way as for dapps?

E.g. a Snap may have permission to do personal_sign already

Reply:
> So given that Snaps may already have eth_accounts permissions on technically all chains, why aren't we granting that permission in the same way as for dapps?

The only reason Snaps have these permissions across all chains (still) is because we haven't added the new chainPermission to snaps yet.

The way the Multichain API works permission-wise is more or less the same as the way the old API works now that we've added chain permissions. Functionally they are obviously quite different: 
- the way you would use permissions for chain X / account 1 on the legacy injected API is you would be switch the wallet to the chain x using that chain permission and then call `eth_sendTransaction` with account 1 using that `eth_accounts` permission
- the way this works on the multichain API is in one atomic call you make an `eth_sendTransaction` request on chain x with account 1, and we validate that CAIP-25 permission for this dapp includes this account on this chain.

Reply:
I understand that, I'm just having trouble following whether the permissions are being migrated correctly.

Is granting `wallet:eip155` the same as granting access to all chains by default? E.g. the old eth_accounts behaviour? If not, then I don't understand fully how the migration works for Snaps.

Reply:
The only reason is that snaps currently do not have permittedChains permissions and so granting permissions to `eip155:<chainId>` scopes would effectively be giving snaps permittedChains permissions. My understanding was that we wanted to hold off on doing this until we are ready to release snap selected networks and chain switching

Reply:
That's fine, but it still doesn't answer my question as to what permission are we then actually granting. What does giving them `wallet:eip155` grant them permission to do?

Reply:
permission to call any EIP-1193 window.ethereum exposed method, incl signing methods for the accounts specified in the `wallet:eip155` scope, on the globally selected network

Reply:
OK sorry obviously I misunderstood your line of questioning here in my previous answer.

> So given that Snaps may already have eth_accounts permissions on technically all chains, why aren't we granting that permission in the same way as for dapps?

I think (?) I understand the question better now.

So the new permission format (modeled closely after CAIP-25 request structure) nests accounts under specific chain permissions:

```
      "eip155:10": {
        "accounts:" ["eip155:10:0x0910e12C68d02B561a34569E1367c9AAb42bd810" 
        ,"eip155:10:0x806627172aF48bD5B0765D3449A7DEf80d6576ff"]
      },
      "eip155:42161": {
        "accounts":["eip155:42161:0x0910e12C68d02B561a34569E1367c9AAb42bd810", "eip155:42161:0x806627172aF48bD5B0765D3449A7DEf80d6576ff"],
      },
```
so if we were to grant permissions to snaps in the same way we would necessarily be constraining these account permissions to the enumerated chains.

Putting snaps account permissions into `wallet:eip155` is a temporary way to get around this entanglement of chain and account permissions. Really in the new permissions structure accounts permissions shouldn't really be separable from chain permissions. 

Now its also the case we currently don't have any cases enabled where any accounts **are not included in all the chain scopes** so, until we do, the way it works is all permissioned accounts are permissioned on all permissioned chains. But this still doesn't lead to as broad of a permission as the legacy `eth_accounts` permission (before chain Permissions) since it was unconstrained by what subset of enabled chains the dapp has been granted. That's why we've carved out this temporary solution for snaps until they have chain permissions as well.